examples/systemcalls/systemcalls.c

```diff
@@ -1,99 +1,177 @@
+
+	va_end(args);
+
+	/*
+	 * TODO:
+	 *   Execute a system command by calling fork, execv(),
+	 *   and wait instead of system (see LSP page 161).
+	 *   Use the command[0] as the full path to the command to execute
+	 *   (first argument to execv), and use the remaining arguments
+	 *   as second argument to the execv() command.
+	 *
+	 */
+
+	pid_t pid = fork();
+
+	if (-1 == pid)
+	{
+		perror("Fork failed : ");
+		return false;
+	}
+	else if (0 == pid)
+	{
+		execv(command[0], command);
+		perror("execv() call failed!");
+		_exit(EXIT_FAILURE);
+	}
+
+	int status;
+	waitpid(pid, &status, 0);
+
+	if (WIFEXITED(status) && EXIT_FAILURE == WEXITSTATUS(status))
+		return false;
+
+	return true;
 }
 
 /**
-* @param outputfile - The full path to the file to write with command output.
-*   This file will be closed at completion of the function call.
-* All other parameters, see do_exec above
-*/
+ * @param outputfile - The full path to the file to write with command output.
+ *   This file will be closed at completion of the function call.
+ * All other parameters, see do_exec above
+ */
 bool do_exec_redirect(const char *outputfile, int count, ...)
 {
-    va_list args;
-    va_start(args, count);
-    char * command[count+1];
-    int i;
-    for(i=0; i<count; i++)
-    {
-        command[i] = va_arg(args, char *);
-    }
-    command[count] = NULL;
-    // this line is to avoid a compile warning before your implementation is complete
-    // and may be removed
-    command[count] = command[count];
-
-
-/*
- * TODO
- *   Call execv, but first using https://stackoverflow.com/a/13784315/1446624 as a refernce,
- *   redirect standard out to a file specified by outputfile.
- *   The rest of the behaviour is same as do_exec()
- *
-*/
-
-    va_end(args);
-
-    return true;
+	va_list args;
+	va_start(args, count);
+	char *command[count + 1];
+	int i;
+	for (i = 0; i < count; i++)
+	{
+		command[i] = va_arg(args, char *);
+	}
+	command[count] = NULL;
+	// this line is to avoid a compile warning before your implementation is complete
+	// and may be removed
+	command[count] = command[count];
+	va_end(args);
+
+	/*
+	 * TODO
+	 *   Call execv, but first using https://stackoverflow.com/a/13784315/1446624 as a refernce,
+	 *   redirect standard out to a file specified by outputfile.
+	 *   The rest of the behaviour is same as do_exec()
+	 *
+	 */
+
+	int fd = open(outputfile, O_WRONLY | O_TRUNC | O_CREAT, 0644);
+	if (0 > fd)
+	{
+		perror("File could not be opened!");
+		return false;
+	}
+
+	pid_t pid = fork();
+
+	if (-1 == pid)
+	{
+		perror("Fork failed : ");
+		close(fd);
+		return false;
+	}
+	else if (0 == pid)
+	{
+		if (dup2(fd, 1) < 0)
+		{
+			perror("dup2() failed!");
+			_exit(EXIT_FAILURE);
+		}
+
+		close(fd);
+		execv(command[0], command);
+		perror("execv() call failed!");
+		_exit(EXIT_FAILURE);
+	}
+
+	int status;
+	waitpid(pid, &status, 0);
+
+	close(fd);
+
+	if (WIFEXITED(status) && EXIT_FAILURE == WEXITSTATUS(status))
```

Comment:
Good strategy for error handling and cleanup.
Are these checks sufficient to ensure any error return code results in a false return value though?  I think you may need to compare to 0 for success instead.